keras_nlp/src/utils/transformers_model_utils/hf_common_port.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import einops
+from safetensors import safe_open
+
+
+def set_keras_weights(
+    safetensor_files,
+    safetensor_config,
+    keras_weight,
+    hf_weight_keys,
+    rearrange_patterns=None,
+    rearrange_dims=None,
+):
+    """
+    Set Keras model weights from SafeTensors file.
+
+    Args:
+        safetensor_files (dict): Dictionary of SafeTensor file paths.
+        safetensor_config (dict): Configuration for SafeTensors.
+        keras_weight (keras.layers.Layer): Keras layer to set the weights for.
+        hf_weight_keys (str or list): Key(s) for the Hugging Face weight(s).
+        rearrange_patterns (str or list, optional): Pattern(s) for rearranging dimensions using einops.
+        rearrange_dims (dict, optional): Dimensions for rearranging using einops.
+    """
+    if isinstance(hf_weight_keys, str):
+        hf_weight_keys = [hf_weight_keys]
+    if rearrange_patterns and isinstance(rearrange_patterns, str):
+        rearrange_patterns = [rearrange_patterns] * len(hf_weight_keys)
+    elif not rearrange_patterns:
+        rearrange_patterns = [None] * len(hf_weight_keys)
+
+    tensors = []
+    for hf_weight_key, rearrange_pattern in zip(
+        hf_weight_keys, rearrange_patterns
+    ):
+        safetensor_file = safetensor_files[
+            safetensor_config["weight_map"][hf_weight_key]
+        ]
+        with safe_open(safetensor_file, framework="np") as f:
+            tensor = f.get_tensor(hf_weight_key)
+            if rearrange_pattern:
+                tensor = einops.rearrange(
```

Comment:
@mattdangerw I have added code to bypass our usage of `einops.rearrange`. Can you review and let me know what you think about it.

Reply:
Seems a little hard to read, particularly with the shapes pass as ints formatted into strings. Alternate proposal...

- We do weight setting at the individual weight level, instead of the entire layer. So `dense.kernel.assign(array)` instead of `dense.set_weights`. This is slightly more efficient in terms of peak memory usage, as you don't need to read in a entire layers worth of weights at once.
- We have either a hook to run an arbitrary function on each loaded numpy array, or just an option to transpose the weight.
- We have an option to reshape to the keras weight shape before assigning.

So something like this
```python
port_weight(
    decoder_layer.attention.key_dense.kernal,
    f"model.layers.{i}.self_attn.k_proj.weight",
    transpose=True,
    reshape=True,
)

# becomes...
weight = f.get_tensor(hf_weight_key)
if transpose:
    np.tranpose(weight)
if reshape:
    np.reshape(weight, ops.shape(keras_variable))
keras_variable.assign(weight)
```

Might be cases I am missing, but overall I think this might be more concise and readable?

Reply:
This is much more readable.

I am trying to understand how this approach could handle something like `einops.rearrange("(a c) b -> a b c", a=8)`. Here we need to reshape the tensor first and then transpose.

Reply:
I really liked the hook idea.

```py
def port_weight(keras_weight, hf_weight_key, hook_fn=None):
    safetensor_file = safetensor_files[hf_safetensor_config["weight_map"][hf_weight_key]]
    with safe_open(safetensor_file, framework="np") as f:
        tensor = f.get_tensor(hf_weight_key)
        if hook_fn:
            tensor = hook_fn(tensor, list(keras_weight.shape))
        np.testing.assert_array_equal(tensor, keras_weight)
```

and we could call the function like this

```py
port_weight(
    keras_weight = keras_backbone.get_layer(f"decoder_block_{i}").attention.query_dense.get_weights()[0],
    hf_weight_key = f"model.layers.{i}.self_attn.q_proj.weight",
    # rearrange_pattern = "(a c) b -> a b c",
    # rearrange_dims = {"a": 8}
    hook_fn = lambda hf_tensor, keras_shape: np.transpose(
        np.reshape(
            hf_tensor,
            (keras_shape[0], keras_shape[2], keras_shape[1])),
        axes=(0, 2, 1)
    )
)
```

What do you think about this?

Reply:
thanks! i think that's readable yeah

Reply:
I have made the changes! Could you please review the changes?

The next steps for me would be to add the tests.